### PR TITLE
Allow configuring `jib` plugin type

### LINKED ISSUE #2963: Jib configuration block should allow explicitly setting type
With #2808 we merged the `jibMaven` and `jibGradle` blocks into a single `jib` block where we determine the plugin-type based on the presence of [key files like `build.gradle`](https://github.com/GoogleContainerTools/skaffold/blob/4145d76cbccd0ea46ee31541f5f550368e257fa3/pkg/skaffold/jib/jib.go#L110-L129).  This isn't fool-proof and we should allow setting the value explicitly.

### PR DESCRIPTION
**Pull request by briandealwis**
Fixes #2963.  

Note that this is a somewhat unorthodox change as it exposes an internal and previously ignored field from the `v1beta14`'s `jib` config.

**Description**

In #2808 we unified the `jibMaven` and `jibGradle` configuration blocks into a single `jib` block.  The Skaffold-Jib builder looks for certain files to determine the Jib plugin type (Maven or Gradle).  We've had a report where Skaffold-Jib was unable to guess the Jib plugin type.  This PR adds the ability to explicitly configure the `jib` block to specify the plugin type.

The unified [`JibArtifact` already has a `Type` field](https://github.com/GoogleContainerTools/skaffold/blob/567a0800792804933941fb44ae8698a8afe1b68e/pkg/skaffold/schema/latest/config.go#L781-L782) which allows explicitly specifying the builder type.  This field is internal and not marshalled.  The intent was to expose this field to encode a Go enum (with corresponding `MarshalYAML()` and `UnmarshalYAML()`) but I punted at the time as it required teaching the schema generator to output [JSON schema `enum`](https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values).  But I realized that the `pkg/skaffold/schema/validation` approach was easier, but it requires changing the type of `Type`.

**User facing changes**
This PR changes the type of `JibArtifact`'s `Type` field and exposes it.  As a result:

- `v1beta14` and `v1beta15`'s `jib` now have optional `type` fields
- `skaffold fix` on `v1beta13` `jibMaven` and `jibGradle` configurations will now have a `type` field

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit test added.
- [ ] User facing changes look good.

**Release Notes**

`jib` artifacts can explicitly specify the plugin type as `maven` or `gradle`

### CODE REVIEW

File: pkg/skaffold/schema/validation/validation.go [hunk 1]
```diff
@@ -187,3 +188,18 @@ func validatePortForwardResources(pfrs []*latest.PortForwardResource) []error {
...
+func validateJibPluginTypes(artifacts []*latest.Artifact) (errs []error) {
+	for _, a := range artifacts {
+		if a.JibArtifact == nil || a.JibArtifact.Type == "" {
+			continue
+		}
+		t := strings.ToLower(a.JibArtifact.Type)
+		if t == "maven" || t == "gradle" {
+			continue
+		}
+		errs = append(errs, fmt.Errorf("artifact %s has invalid plugin type '%s'", a.ImageName, t))
```
Comment (chanseokoh) [inline pkg/skaffold/schema/validation/validation.go]: Is it possible that `a.ImageName` is empty?

"... has invalid Jib plugin type..." ?

File: pkg/skaffold/schema/validation/validation.go [hunk 2]
```diff
@@ -187,3 +188,18 @@ func validatePortForwardResources(pfrs []*latest.PortForwardResource) []error {
...
+	}
+	return
```
Comment (chanseokoh) [inline pkg/skaffold/schema/validation/validation.go]: `return errs`

Comment (briandealwis) [reply pkg/skaffold/schema/validation/validation.go]: It's a named return variable (this was a copy-n-pasta of `validateDockerNetworkMode` above

File: pkg/skaffold/schema/validation/validation.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/schema/validation/validation.go]: Image names are validated further above in `validateImageNames()`.  There's not much we can do at this point.

File: pkg/skaffold/schema/latest/config.go [hunk 3]
```diff
@@ -778,6 +778,8 @@ type JibArtifact struct {
 	// For example: `["--no-build-cache"]`.
 	Flags []string `yaml:"args,omitempty"`
 
-	// Type the Jib builder type (internal: see jib.PluginType)
-	Type int `yaml:"-"`
```
Comment (balopat) [inline pkg/skaffold/schema/latest/config.go]: This should be fine as it actually helps making the upgrade process to not break when the detection mechanism doesn't work. It is a conservative change. I will use admin rights to merge this PR.

**Comment by briandealwis** [comment]
Sorry @balopat, this totally slipped through the cracks.

**Comment by briandealwis** [comment]
I need to update this in light of #3041

**Comment by briandealwis** [comment]
I hacked up the schema generator to emit JSON schema `enum`s.  PTAL!

**Comment by balopat** [comment]
This is awesome, thanks for putting the enum together as well!

**Comment by balopat** [comment]
Merged with admin privileges for the historical version change as it actually helps with upgrades.

**Pull request merged**